android/lib/src/main/java/com/swmansion/gesturehandler/GestureHandler.kt

```diff
@@ -0,0 +1,482 @@
+package com.swmansion.gesturehandler
+
+import android.view.MotionEvent
+import android.view.MotionEvent.PointerCoords
+import android.view.MotionEvent.PointerProperties
+import android.view.View
+import com.facebook.react.bridge.UiThreadUtil
+import java.util.*
+
+open class GestureHandler<ConcreteGestureHandlerT : GestureHandler<ConcreteGestureHandlerT>> {
+  private val mTrackedPointerIDs = IntArray(MAX_POINTERS_COUNT)
+  private var mTrackedPointersCount = 0
+  var tag = 0
+  var view: View? = null
+    private set
+  var state = STATE_UNDETERMINED
+    private set
+  var x = 0f
+    private set
+  var y = 0f
+    private set
+  var isWithinBounds = false
+    private set
+  var isEnabled = true
+    private set
+
+  private var mHitSlop: FloatArray? = null
+  var eventCoalescingKey: Short = 0
+    private set
+  var lastAbsolutePositionX = 0f
+    private set
+  var lastAbsolutePositionY = 0f
+    private set
+
+  private var mLastEventOffsetX = 0f
+  private var mLastEventOffsetY = 0f
+  private var mShouldCancelWhenOutside = false
+  var numberOfPointers = 0
+    private set
+  private var mOrchestrator: GestureHandlerOrchestrator? = null
+  private var mListener: OnTouchEventListener<ConcreteGestureHandlerT>? = null
+  private var mInteractionController: GestureHandlerInteractionController? = null
+
+  @Suppress("UNCHECKED_CAST")
+  protected fun self(): ConcreteGestureHandlerT = this as ConcreteGestureHandlerT
+
+  protected inline fun applySelf(block: ConcreteGestureHandlerT.() -> Unit): ConcreteGestureHandlerT =
+    self().apply { block() }
+
+  // set and accessed only by the orchestrator
+  @JvmField
+  var mActivationIndex = 0
+
+  // set and accessed only by the orchestrator
+  @JvmField
+  var mIsActive = false
+
+  // set and accessed only by the orchestrator
+  @JvmField
+  var mIsAwaiting = false
+
+  open fun dispatchStateChange(newState: Int, prevState: Int) {
+    mListener?.onStateChange(self(), newState, prevState)
+  }
+
+  open fun dispatchTouchEvent(event: MotionEvent?) {
+    mListener?.onTouchEvent(self(), event)
+  }
+
+  open fun resetConfig() {
+    mShouldCancelWhenOutside = false
+    isEnabled = true
+    mHitSlop = null
+  }
+
+  fun hasCommonPointers(other: GestureHandler<*>): Boolean {
+    for (i in mTrackedPointerIDs.indices) {
+      if (mTrackedPointerIDs[i] != -1 && other.mTrackedPointerIDs[i] != -1) {
+        return true
+      }
+    }
+    return false
+  }
+
+  fun setShouldCancelWhenOutside(shouldCancelWhenOutside: Boolean) = applySelf {
```

Comment:
For me it was not instantly obvious that this method (and others using `applySelf`) is returning a `ConcreteGestureHandlerT`. Maybe we should state is explicitly?
```suggestion
  fun setShouldCancelWhenOutside(shouldCancelWhenOutside: Boolean): ConcreteGestureHandlerT
    = applySelf {
```

Reply:
Resolved in b565f2f.